File: Data Structures & Algorithms/binary-tree-diameter/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        left = self.getHeight(root.left)
        right = self.getHeight(root.right)

        return max(left + right, self.diameterOfBinaryTree(root.left), self.diameterOfBinaryTree(root.right))

    def getHeight(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        return 1 + max(self.getHeight(root.left), self.getHeight(root.right))

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-3.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        val queue = ArrayDeque<TreeNode>()
        var level = 0
        val result = mutableListOf<List<Int>>()

        if (root != null) {
            queue.add(root)
        }

        while (queue.size > 0) {
            val size = queue.size
            var innerResult = mutableListOf<Int>()
            for (i in 0 until size) {
                val current = queue.removeFirst()
                innerResult.add(current.`val`)
                current.left?.let { queue.add(it) }
                current.right?.let { queue.add(it) }
            }
            result.add(innerResult)
            level++
        }

        return result
    }
}
